pkg/util/virtualmachine/virtualmachine.go

```diff
@@ -50,3 +51,20 @@ func IsVMStopped(
 
 	return false, nil
 }
+
+// ListByInstanceLabels gets a list of VMs whose instance labels match the
+// specified labels.
+func ListByInstanceLabels(namespace string, selector labels.Selector, cache ctlkubevirtv1.VirtualMachineCache) ([]*kubevirtv1.VirtualMachine, error) {
+	vmList, err := cache.List(namespace, labels.Everything())
+	if err != nil {
+		return nil, err
+	}
+	vmMatchList := make([]*kubevirtv1.VirtualMachine, 0, len(vmList))
+	for _, vm := range vmList {
+		if !selector.Matches(labels.Set(vm.Spec.Template.ObjectMeta.Labels)) {
```

Comment:
Why not set label on VM directly, so we don't need to search the label by ourselves? Also, original design takes O(N) time complexity to search all VMs in the namespace.

BTW, if we design the label on VMI, users need to restart VM to let it take effect.

https://github.com/harvester/harvester/assets/4927361/2391270c-4030-4867-895d-c2ee4392a1ff

Reply:
I'm currently adapting the code which will not use instance labels any more. Making sure that my special label is copied to VMI justifies the effort to then have the option of setting the label using kubectl and filters for the desired VMs.

> BTW, if we design the label on VMI, users need to restart VM to let it take effect.

I think this is not right because the labels are always synced on changes on the VM.
- https://github.com/harvester/harvester/blob/f9464f06692f046f07b5bd5d9cbea3ed182c1610/pkg/controller/master/virtualmachine/register.go#L67
- https://github.com/harvester/harvester/blob/f9464f06692f046f07b5bd5d9cbea3ed182c1610/pkg/controller/master/virtualmachine/vm_controller.go#L168